### PR TITLE
Core/feature mpi flag synchronization

### LINKED ISSUE #1277: MPI flags synchronize
After a short discussion with @jcotela we realized that still not have any convention about the flag synchronization in MPI. As a preliminary idea, we think that we could add the next two methods:

- SynchronizeFlagAnd()
- SynchronizeFlagOr()

that perform the and and or boolean operations respectively. 

Any suggestion/idea about that?

### PR DESCRIPTION
**Pull request by jcotela**
Closes #1277.

This is only a partial implementation, but I am creating the PR to check if appveyor can compile it. I will update the PR once this is ready to merge.

Marking as behavior change because I am modifying the `&=` operator of `Kratos::Flags`.

### CODE REVIEW

**Comment by philbucher** [comment]
I think this time appveyor had an actual error, nothing related to time-out

**Comment by jcotela** [comment]
@philbucher I am aware of the appveyor error and it is related to having divided flags.h into header and source. I will revert it before the final PR, but I am keeping it for now since it lets me play with that class without having to recompile kratos every time.

**Comment by jcotela** [comment]
Note: this PR now requires #4316

**Comment by jcotela** [comment]
Thanks to @roigcarlo, who fixed the windows compilation issues, this is ready for review now.

This PR closes #1277.

File: kratos/CMakeLists.txt [hunk 1]
```diff
@@ -39,7 +39,6 @@ set( KRATOS_CORE_SOURCES
     ${CMAKE_CURRENT_SOURCE_DIR}/sources/standard_linear_solver_factory.cpp;
     ${CMAKE_CURRENT_SOURCE_DIR}/sources/standard_preconditioner_factory.cpp;
     #${CMAKE_CURRENT_SOURCE_DIR}/sources/kratos_exception.cpp;
-    ${CMAKE_CURRENT_SOURCE_DIR}/containers/flags.cpp;
```
Comment (pooyan-dadvand) [inline kratos/CMakeLists.txt]: So the flags.cpp is removed from cmake but not from the folder?

**Comment by pooyan-dadvand** [comment]
What was the problem in windows?

File: kratos/CMakeLists.txt (hunk 1 above)
Comment (roigcarlo) [reply kratos/CMakeLists.txt]: This file is not changed anymore, I think you are comparing and outdated version :S

**Comment by roigcarlo** [comment]
@pooyan-dadvand  the problem is that binary operators where not exported. Since they need to be defined outside the class but we use them inside they needed to be friends of the class and seems that in that particular situation the `__declspec(dllexport)` needs to be present in the friend declaration even when the class is exported itself.

**Comment by pooyan-dadvand** [comment]
I see, This is normal as these methods are not belong to the class so are not exposed. What is important is to maintain them as friend.

**Comment by jcotela** [comment]
@pooyan-dadvand ~could you please check 122332be218520187938d617bc486e0ebb22c779 in particular?~ I redefined the `Set` operation to work on multiple flags at the same time (now it essentially overwrites the bits defined by the rhs), so now things like `node.Set(INLET | STRUCTURE, true)` should work out of the box (note that it is equivalent to `node.Set(INLET & STRUCTURE, true)`, which does the same thing...)

**Edit**: looks like I referenced the wrong commit, the one I am speaking about above is eb48c29210859b9788235d1037e60e8cb203ce42

File: kratos/tests/cpp_tests/containers/test_flags.cpp [hunk 2]
```diff
@@ -119,14 +119,11 @@ KRATOS_TEST_CASE_IN_SUITE(KratosFlagsSetMultiple, KratosCoreFastSuite) {
...
 
-    KRATOS_CHECK_EQUAL( flags1 == flags2, false );
-    KRATOS_CHECK_EQUAL( flags1 != flags2, true );
-
-    KRATOS_CHECK_EQUAL( (!flags1) == flags2, true );
-    KRATOS_CHECK_EQUAL( flags1 == !flags2, true );
+    KRATOS_CHECK_EQUAL( INLET == NOT_INLET, false );
+    KRATOS_CHECK_EQUAL( INLET != NOT_INLET, true );
+    // trivial check (it is acutally 0 == 0) just to see that everything works when the first argument of == is not a lhs
+    KRATOS_CHECK_EQUAL( (~INLET & INLET) == NOT_INLET, true );
```
Comment (pooyan-dadvand) [inline kratos/tests/cpp_tests/containers/test_flags.cpp:126]: This result of `(~INLET & INLET)` shouldn't be zero?

File: kratos/tests/cpp_tests/containers/test_flags.cpp [hunk 3]
```diff
@@ -141,13 +138,30 @@ KRATOS_TEST_CASE_IN_SUITE(KratosFlagsOperators, KratosCoreFastSuite) {
...
 
+    flags_and.Clear();
+    flags_and = (~flags2) & flags1;
+
+    KRATOS_CHECK_EQUAL(flags_and.IsDefined(INLET), true);
+    KRATOS_CHECK_EQUAL(flags_and.IsDefined(OUTLET), true);
+    KRATOS_CHECK_EQUAL(flags_and.Is(INLET), true);
+    KRATOS_CHECK_EQUAL(flags_and.Is(OUTLET), false);
+
     Kratos::Flags flags3(NOT_OUTLET | VISITED);
```
Comment (pooyan-dadvand) [inline kratos/tests/cpp_tests/containers/test_flags.cpp:165]: we should pass this test as logic exam to the new developers!! 😄

**Review by pooyan-dadvand — COMMENTED** [review]
Everything is ok, but my only concern is `(~INLET & INLET)` which intuitively should be zero.

File: kratos/tests/cpp_tests/containers/test_flags.cpp (hunk 2 above)
Comment (jcotela) [reply kratos/tests/cpp_tests/containers/test_flags.cpp:126]: zero with the inlet flag defined :man_facepalming:

**Review by pooyan-dadvand — APPROVED** [review]
Now I see why the result was correct! Thanks!

**Pull request merged**